src/tests/test_dal_middleware.py

```diff
@@ -0,0 +1,303 @@
+    for i in range(0, len(expected_middlewares)):
+        assert isinstance(dal._reversed_middleware[0], type(expected_middlewares[0]))
+
+
+def test_dal_method_middleware():
+    """
+    Verify that when calling a method on a service off the DAL,
+    the method middlewares run, in the correct order and run
+    before the method itself is called.
+
+
+    We deliberately do not use the DalMethodResolverMiddleware here,
+    because we are testing middleware call chains, not functionality
+    of that class.
+    """
+    middleware_requests = []
+
+    def test_method():
+        middleware_requests.append(('default-handler-call', None))
+
+    class OuterMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('outer-middleware-ingress', request))
+            result = handler(request)
+            middleware_requests.append(('outer-middleware-egress', request))
+            return result
+
+    class SecondMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('second-middleware-ingress', request))
+            result = handler(request)
+            middleware_requests.append(('second-middleware-egress', request))
+            return result
+
+    # Verify custom default middleware
+    class DefaultMiddleware(MethodMiddleware):
+        def __call__(self, request, handler):
+            middleware_requests.append(('default-middleware-ingress', request))
+            # We need to get past the DalMethodResolverMiddleware
+            # which sets the `dal_method` attribute to
+            # the method that has to be called
+            request.dal_method = test_method
+            result = handler(request)
+            middleware_requests.append(('default-middleware-egress', request))
+            return result
+
+    dm = DataManager()
+
+    dal = DataAccessLayer(data_manager=dm, middleware=[OuterMiddleware, SecondMiddleware],
+                          default_middleware=(DefaultMiddleware,))
+
+    with dm.context():
+        dal.fake.service.method()
+
+    first_middleware_to_run = middleware_requests[0]
+    assert first_middleware_to_run[0] == 'outer-middleware-ingress'
+    second_middleware_to_run = middleware_requests[1]
+    assert second_middleware_to_run[0] == 'second-middleware-ingress'
+    default_middleware_to_run = middleware_requests[2]
+    assert default_middleware_to_run[0] == 'default-middleware-ingress'
+
+    default_method_to_run = middleware_requests[3]
+    assert default_method_to_run[0] == 'default-handler-call'
+
+    default_middleware_to_run_egress = middleware_requests[4]
+    assert default_middleware_to_run_egress[0] == 'default-middleware-egress'
+    second_middleware_to_run_egress = middleware_requests[5]
+    assert second_middleware_to_run_egress[0] == 'second-middleware-egress'
+    first_middleware_to_run_egress = middleware_requests[6]
```

Comment:
What if there was a `middleware_requests[7]` or higher and you didn't know it?

- [ ] Verify complete response or verify expected length